### 0. 파일 불러오기

In [1]:
import json
import pandas as pd

#### 0-1. prediction.json

In [2]:
with open('sparse_prediction.json') as f: #⭐
    sparse_ensemble_json = json.loads(f.read())
sparse_ensemble_df = pd.DataFrame(sparse_ensemble_json,  index=['sparse'])
sparse_ensemble_df = sparse_ensemble_df.transpose()

with open('ssm_prediction.json') as f: #⭐
    ssm_ensemble_json = json.loads(f.read())
ssm_ensemble_df = pd.DataFrame(ssm_ensemble_json,  index=['sparse'])
ssm_ensemble_df = ssm_ensemble_df.transpose()

with open('hybrid_prediction.json') as f:
    hybrid_ensemble_json = json.loads(f.read())
hybrid_ensemble_df = pd.DataFrame(hybrid_ensemble_json,  index=['sparse'])
hybrid_ensemble_df = hybrid_ensemble_df.transpose()

with open('electra_prediction.json') as f:
    electra_ensemble_json = json.loads(f.read())
electra_ensemble_df = pd.DataFrame(electra_ensemble_json,  index=['sparse'])
electra_ensemble_df = electra_ensemble_df.transpose()

with open('predictions_dense.json') as f:
    dense_ensemble_json = json.loads(f.read())
dense_ensemble_df = pd.DataFrame(dense_ensemble_json,  index=['sparse'])
dense_ensemble_df = dense_ensemble_df.transpose()

#### 0-2. nbest_prediction.json

In [3]:
arr = ['1', '2','3','4','5','6','7','8','9','10', '11','12','13','14','15','16','17','18','19','20']
with open('nbest_predictions_hybrid.json') as f:
    hybrid_json = json.loads(f.read())
hybrid_df = pd.DataFrame(hybrid_json,  index=arr)
hybrid_df = hybrid_df.transpose()

In [4]:
with open('nbest_predictions_sparse.json') as f: 
    sparse_json = json.loads(f.read())
sparse_df = pd.DataFrame(sparse_json,  index=arr)
sparse_df = sparse_df.transpose()
#sparse_df

In [5]:
with open('nbest_predictions_electra.json') as f:
    electra_json = json.loads(f.read())
electra_df = pd.DataFrame(electra_json,  index=arr)
electra_df = electra_df.transpose()
#electra_df

In [6]:
with open('nbest_predictions_ssm.json') as f:
    ssm_json = json.loads(f.read())
ssm_df = pd.DataFrame(ssm_json,  index=arr)
ssm_df = ssm_df.transpose()
#electra_df

In [7]:
with open('nbest_predictions_dense.json') as f:
    dense_json = json.loads(f.read())
dense_df = pd.DataFrame(dense_json,  index=arr)
dense_df = dense_df.transpose()

### 1. 600개 질문에 대한 max_prob list로 표현

In [8]:
hybrid_df.iloc[0][1]['probability'] # iloc 사용법

0.36767578125

In [9]:
hybrid_max_probs = []
for i in range(600):
    tmp = 0
    for j in range(20):
        tmp = max(tmp, hybrid_df.iloc[i][j]['probability'])
    hybrid_max_probs.append(tmp)
#print(hybrid_max_probs[0])

In [10]:
sparse_max_probs = []
for i in range(600):
    tmp = 0
    for j in range(20):
        tmp = max(tmp, sparse_df.iloc[i][j]['probability'])
    sparse_max_probs.append(tmp)
#print(sparse_max_probs[0])

In [11]:
electra_max_probs = []
for i in range(600):
    tmp = 0
    for j in range(20):
        tmp = max(tmp, electra_df.iloc[i][j]['probability'])
    electra_max_probs.append(tmp)
#print(electra_max_probs[0])

In [12]:
ssm_max_probs = []
for i in range(600):
    tmp = 0
    for j in range(20):
        tmp = max(tmp, ssm_df.iloc[i][j]['probability'])
    ssm_max_probs.append(tmp)
#print(ssm_max_probs[0])

In [13]:
dense_max_probs = []
for i in range(600):
    tmp = 0
    for j in range(20):
        tmp = max(tmp, ssm_df.iloc[i][j]['probability'])
    dense_max_probs.append(tmp)

### 2. 기준값 찾기 위해 sparse max_prob 분석하기 ‼

In [14]:
average = 0
for i in range(600):
    average += sparse_max_probs[i] 
print("Average: ",average/600)
print("Min: ", min(sparse_max_probs))
print("Max: ", max(sparse_max_probs))

Average:  0.8167743937174479
Min:  0.12017822265625
Max:  1.0


In [15]:
cnt = 0
for i in range(600):
    if sparse_max_probs[i] < 0.5:
        cnt +=1
print(cnt)

54


*sparse*
probabiltiy가 0.8보다 작은 답안 228개 (38%) <br>
probabiltiy가 0.5보다 작은 답안 54개 (9%)<br>
다른 답안으로 교체되어도 괜찮지 않을까 <br>

#### hybrid 분석하기

In [16]:
average = 0
for i in range(600):
    average += hybrid_max_probs[i] 
print("Average: ",average/600)
print("Min: ", min(hybrid_max_probs))
print("Max: ", max(hybrid_max_probs))

cnt = 0
for i in range(600):
    if hybrid_max_probs[i] < 0.7:
        cnt +=1
print(cnt)

Average:  0.818145039876302
Min:  0.10064697265625
Max:  1.0
175


*hybrid*
probabiltiy가 0.8보다 작은 답안 224개 (37%)<br>
probabiltiy가 0.7보다 작은 답안 175개 (29%)<br>
probabiltiy가 0.5보다 작은 답안 44개<br>
다른 답안으로 교체되어도 괜찮지 않을까 

#### electra 분석하기

In [17]:
average = 0
for i in range(600):
    average += electra_max_probs[i] 
print("Average: ",average/600)
print("Min: ", min(electra_max_probs))
print("Max: ", max(electra_max_probs))

cnt = 0
for i in range(600):
    if electra_max_probs[i] < 0.5:
        cnt +=1
print(cnt)

Average:  0.7965768725052476
Min:  0.07231848686933517
Max:  0.9999853372573853
65


*electra*
probabiltiy가 0.8보다 작은 답안 248개
probabiltiy가 0.7보다 작은 답안 199개
probabiltiy가 0.5보다 작은 답안 65개
다른 답안으로 교체되어도 괜찮지 않을까 

#### ssm 분석하기 ‼

In [18]:
average = 0
for i in range(600):
    average += ssm_max_probs[i] 
print("Average: ",average/600)
print("Min: ", min(ssm_max_probs))
print("Max: ", max(ssm_max_probs))

cnt = 0
for i in range(600):
    if ssm_max_probs[i] < 0.8:
        cnt +=1
print(cnt)

Average:  0.8135212089618047
Min:  0.15367691218852997
Max:  0.999993085861206
232


#### sparse + hybrid 분석하기 (둘이 답변이 0.89로 가장 유사하므로 확인차)

In [19]:
sparse_hybrid = []
for i in range(600):
    middle = (sparse_max_probs[i] + hybrid_max_probs[i])/2
    sparse_hybrid.append(middle)
#print(sparse_hybrid)

average = 0
for i in range(600):
    average += sparse_hybrid[i] 
print("Average: ",average/600)
print("Min: ", min(sparse_hybrid))
print("Max: ", max(sparse_hybrid))

Average:  0.817459716796875
Min:  0.11041259765625
Max:  1.0


## 앙상블
### 기준 1. sparse, hybrid 평균이 0.35보다 작은 거 ssm으로 교체

In [20]:
cnt = 0
idx = []
for i in range(600):
    if sparse_hybrid[i] < 0.35:
        cnt +=1
        idx.append(i)
print(cnt)

11


0.7로하니까 네페로스 행성부터 바뀌는 문제 발생

In [21]:
print(idx) # sparse가 낮아서 교체할 애들

[25, 28, 51, 112, 134, 170, 327, 330, 473, 530, 595]


In [22]:
for i in idx:
    print(sparse_ensemble_df.iloc[i]['sparse'], '/', ssm_ensemble_df.iloc[i]['sparse'])

관우 / 진경동(秦慶童)
괴팅겐 / 신시내티대학교
미국 대통령 우드로 윌슨 / 미국 대통령 우드로 윌슨
보안 / 보안
외기권의 형성 원리는 무엇인가? / 잠재적 심판의 대상
여름철 / 여름철
종횡무진 한국경제 / <<종횡무진 한국경제>>
1773년 / 1773년
빛 / 빛
LdB 말뫼 / 스타베크 포트발 크빈네르
슈라우드(=소노자키 후미네) / 재단 X


### 기준 2. sparse보다 0.35점 이상 높은 ssm은 ssm을 따르게

In [23]:
ssm_great = []
for i in range(600):
    if (sparse_max_probs[i] + 0.35 <= ssm_max_probs[i]) and (sparse_ensemble_df.iloc[i]['sparse'] != ssm_ensemble_df.iloc[i]['sparse']):
        ssm_great.append(i) # ssm이 높아서 교체할 애들
        
for i in ssm_great:
    print(sparse_ensemble_df.iloc[i]['sparse'], '/', ssm_ensemble_df.iloc[i]['sparse'])

쇼군가와 / 아시카가 쇼군가와
아라라기 / 아라라기 코요미와 오시노 메메
관우 / 진경동(秦慶童)
청법대 / 일등바위
구내 건널목 / 건늠선
1584년 / 1584년(선조 17년)
마리아나 제도에서 벗어나자 / 태풍 초이완이 크게 발달
호해 / 전한 고제
연합국 / 프랑스
풍력이나 조력 발전 등 대체 에너지의 개발 / 대체 에너지의 개발
페니키아 문자 / 키릴 문자
성 베드로 대성전 / 파비아
평산 / 르네상스 양식
서울시 / 부산시
시캉브르 / 잃어버린 세계 건물군
4칸 / 곳간채
1988년 / 라트비아
1649년 / 1722년
비올라 / 콘트라베이스
공화당원 투표인들 만을 화나게 할 것 / 공화당원 투표인들 만을 화나게
슈라우드(=소노자키 후미네) / 재단 X


In [24]:
total_idx = list(set(idx + ssm_great))
print(len(total_idx))
print(total_idx)

for i in total_idx:
    ssm_ensemble_df.iloc[i]['sparse'] = ssm_ensemble_df.iloc[i]['sparse'] 
ssm_ensemble_df

ssm_ensemble_df.to_json('only_ssm_ensemble.json',force_ascii=False)

30
[134, 7, 262, 390, 14, 530, 25, 28, 29, 540, 31, 291, 170, 555, 51, 52, 573, 193, 327, 585, 330, 332, 595, 85, 88, 473, 357, 230, 112, 379]


### 기준 3. sparse, hybrid는 같은데 ssm만 다른 답변 교체 (0.3 이상)

In [25]:
follow_ssm = []
for i in range(600):
    if ((sparse_max_probs[i] + hybrid_max_probs[i])/2 +0.3 <= ssm_max_probs[i]) and (sparse_ensemble_df.iloc[i]['sparse'] == hybrid_ensemble_df.iloc[i]['sparse'] ) and (sparse_ensemble_df.iloc[i]['sparse'] != ssm_ensemble_df.iloc[i]['sparse']):
        follow_ssm.append(i) # ssm이 높아서 교체할 애들
        
for i in follow_ssm:
    print(sparse_ensemble_df.iloc[i]['sparse'], '/', ssm_ensemble_df.iloc[i]['sparse'])

쇼군가와 / 아시카가 쇼군가와
아라라기 / 아라라기 코요미와 오시노 메메
청법대 / 일등바위
구내 건널목 / 건늠선
1584년 / 1584년(선조 17년)
호해 / 전한 고제
연합국 / 프랑스
3년 / 150여 년
풍력이나 조력 발전 등 대체 에너지의 개발 / 대체 에너지의 개발
성 베드로 대성전 / 파비아
서재필 / 언더우드(Underwood)박사
언카스 / 언카스(또는 다운카스, Uncas)
서울시 / 부산시
우윤근 원내대표 / 박근혜 대통령
1988년 / 라트비아
비올라 / 콘트라베이스
공화당원 투표인들 만을 화나게 할 것 / 공화당원 투표인들 만을 화나게
슈라우드(=소노자키 후미네) / 재단 X


In [26]:
print(idx)
print(ssm_great)
print(follow_ssm)

[25, 28, 51, 112, 134, 170, 327, 330, 473, 530, 595]
[7, 14, 25, 29, 31, 52, 85, 88, 193, 230, 262, 291, 332, 357, 379, 390, 540, 555, 573, 585, 595]
[7, 14, 29, 31, 52, 88, 193, 216, 230, 291, 298, 338, 357, 424, 540, 573, 585, 595]


### 기준 4. electra 점수 높은 거 반영

In [27]:
electra_great = []
for i in range(600):
    if (sparse_max_probs[i] + 0.4 <= electra_max_probs[i]) and (sparse_ensemble_df.iloc[i]['sparse'] != electra_ensemble_df.iloc[i]['sparse']):
        electra_great.append(i) # ssm이 높아서 교체할 애들

print(electra_great)
print()
for i in electra_great:
    print(sparse_ensemble_df.iloc[i]['sparse'], '/', electra_ensemble_df.iloc[i]['sparse'])

[30, 178, 295, 395, 422, 526, 538, 555, 583]

신사용 양품 가게 / 증권 중개사
1475년 / 1923년
따뜻하고 건조한 바람 / 구름
공민현 / 다카하시 요시노부
2013년 / 200만 달러
OekakiBBS / 교황 요한 바오로 2세이
조지 워싱턴 대통령 / 표트르 1세
1649년 / 1670년
원 / 백루검


### 기준 5. Dense 점수 높은 거 반영

In [28]:
d_great = []
for i in range(600):
    if (sparse_max_probs[i] + 0.4 <= dense_max_probs[i]) and (sparse_ensemble_df.iloc[i]['sparse'] != dense_ensemble_df.iloc[i]['sparse']):
        d_great.append(i) # ssm이 높아서 교체할 애들

print(d_great)
print()
for i in d_great:
    print(sparse_ensemble_df.iloc[i]['sparse'], '/', dense_ensemble_df.iloc[i]['sparse'])

[14, 25, 41, 52, 85, 153, 193, 262, 332, 351, 390, 433, 464, 480, 552, 555, 595]

아라라기 / 샬롯 엘레노어
관우 / 화흡
잡곡 보리밥 / 은행나무
1584년 / 선조
마리아나 제도에서 벗어나자 / 2019년 5월
영국 / 미
연합국 / 유엔군과 대한민국
페니키아 문자 / 키릴 문자
평산 / 르네상스 양식
경시청 수사1과 / 히로시마 도요 카프
4칸 / 6칸
드주드주아나의 동굴 / 가성 소다 공장
앙리 블랑슈 / 나폴레옹
베다는 / 골룸바노
현무암 화방벽 / 돔
1649년 / 1726년
슈라우드(=소노자키 후미네) / 클러스터


### 파일로 생성

In [29]:
total_idx = list(set(idx + ssm_great + follow_ssm))
print(len(total_idx))
print(total_idx)

34
[134, 7, 262, 390, 14, 530, 25, 28, 29, 540, 31, 291, 424, 170, 555, 298, 51, 52, 573, 193, 327, 585, 330, 332, 338, 595, 85, 88, 473, 216, 357, 230, 112, 379]


In [30]:
for i in total_idx:
    sparse_ensemble_df.iloc[i]['sparse'] = ssm_ensemble_df.iloc[i]['sparse'] 
sparse_ensemble_df

sparse
mrc-1-000653        네페로스 행성
mrc-1-001113          냉전 종식
mrc-0-002191  대통령인 빌헬름 미클라스
mrc-0-003951           뉴질랜드
mrc-1-001272            프랑스
...                     ...
mrc-0-002989           재단 X
mrc-0-001804       자유당계 인사들
mrc-0-003411       서스펜디드 게임
mrc-0-003436          디에프 항
mrc-0-002605            콘드률

[600 rows x 1 columns]

In [31]:
sparse_ensemble_df.to_json('ensemble.json',force_ascii=False)